aws/resource_aws_emr_cluster.go

```diff
@@ -661,13 +677,24 @@ func resourceAwsEMRClusterCreate(d *schema.ResourceData, meta interface{}) error
 			InstanceCount: aws.Int64(int64(m["instance_count"].(int))),
 			InstanceRole:  aws.String(emr.InstanceRoleTypeMaster),
 			InstanceType:  aws.String(m["instance_type"].(string)),
-			Market:        aws.String(emr.MarketTypeOnDemand),
 			Name:          aws.String(m["name"].(string)),
 		}
 
-		if v, ok := m["bid_price"]; ok && v.(string) != "" {
-			instanceGroup.BidPrice = aws.String(v.(string))
-			instanceGroup.Market = aws.String(emr.MarketTypeSpot)
+		if Market, ok := m["market"]; ok {
+			if Market == "ON_DEMAND" {
+				instanceGroup.Market = aws.String(emr.MarketTypeOnDemand)
+			} else if Market == "SPOT" {
+				instanceGroup.Market = aws.String(emr.MarketTypeSpot)
+				if bidPrice, ok := m["bid_price"]; ok {
+					if bidPrice != "" {
+						instanceGroup.BidPrice = aws.String(bidPrice.(string))
+					} else {
+						log.Printf("[INFO] the maximum Spot price is set equal to the On-Demand price")
+					}
+				}
+			} else {
+				return fmt.Errorf("Market parameter must be ON_DEMAND or SPOT not: %s", Market)
```

Comment:
This apply-time logic should be removed in preference of validating this at plan-time by adding the following to the new `market` arguments in the schema:

```go
ValidateFunc: validation.StringInSlice([]string{
	emr.MarketTypeOnDemand,
	emr.MarketTypeSpot,
}),
```

Same applies to `core_instance_group` changes below.